### PR TITLE
add packet parsing for v2 binary format

### PR DESCRIPTION
**Pull request by rogpeppe**
This adds low level functions for parsing the binary
format defined in https://github.com/rescrv/libmacaroons/blob/master/doc/format.txt

### CODE REVIEW

**Comment by mhilton** [comment]
:+1:

**Comment by mattyw** [comment]
:+1:

File: packet-v2.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			return nil, nil, fmt.Errorf("fields out of order")
+		}
+		packets = append(packets, p)
+		prevFieldType = p.fieldType
+		data = rest
+	}
+}
+
+func parsePacketV2(data []byte) ([]byte, packetV2, error) {
+	data, ft, err := parseVarint(data)
```
Comment (alesstimec) [inline packet-v2.go:62]: it worries me a bit that you depend on the number of read bytes to determine the field type..  (on L59)

File: packet-v2.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	if err != nil {
```
Comment (alesstimec) [inline packet-v2.go:63]: should we also check here, if the ft > len(data)?

File: packet-v2.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		}
+		rest, p, err := parsePacketV2(data)
+		if err != nil {
+			return nil, nil, err
+		}
+		if p.fieldType == fieldEOS {
+			return rest, packets, nil
+		}
+		if p.fieldType <= prevFieldType {
+			return nil, nil, fmt.Errorf("fields out of order")
```
Comment (alesstimec) [inline packet-v2.go:45]: can't we simply sort fields instead of returning an error for out-of-order fields?

**Comment by alesstimec** [comment]
a few comments, otherwise LGTM

File: packet-v2.go (hunk 1 above)
Comment (rogpeppe) [inline packet-v2.go:62]: Not as far as I can see. The second value returned by parseVarint is the number we've read, not the number of read bytes.

File: packet-v2.go (hunk 2 above)
Comment (rogpeppe) [inline packet-v2.go:63]: ft is the field type, not a byte count.

File: packet-v2.go (hunk 3 above)
Comment (rogpeppe) [inline packet-v2.go:45]: The grammar becomes more complex if fields can be out of order. Also, the
in-order requirement makes the order of signature calculation more clear.
In short, I think having a single canonical field order is a Good Thing :)

File: packet-v2.go (hunk 1 above)
Comment (alesstimec) [inline packet-v2.go:62]: you are treating it differently below on L64...

Comment (rogpeppe) [inline packet-v2.go:62]: actually i'm not. the data length is encoded as a varint.

**Pull request merged**